File: week2/ex16.R
library(MASS)  # For generating multivariate normal samples
library(ggplot2)
library(gridExtra)
library(nortest)

# Parameters
rho <- 0.5
cov <- matrix(c(1, rho, rho, 1), nrow = 2)
mean <- c(0, 0)
N_samples <- 1000

# Generate samples
set.seed(123)  # for reproducibility
samples <- mvrnorm(n = N_samples, mu = mean, Sigma = cov)

# Define projection vectors
v1 <- c(1, 0)  # x-axis
v2 <- c(0, 1)  # y-axis
v3 <- c(1, 1) / sqrt(2)  # 45-degree line
vs <- list(v1, v2, v3)

# Scatterplot of samples
p1 <- ggplot(data = as.data.frame(samples), aes(x = samples[,1], y = samples[,2])) +
  geom_point() +
  ggtitle('Gaussian random samples') +
  theme(plot.title = element_text(hjust = 0.5))

# 3 histograms of different projections
histograms <- lapply(1:length(vs), function(i) {
  projected_samples <- as.vector(samples %*% vs[[i]])
  p <- ggplot(data = as.data.frame(projected_samples), aes(x = projected_samples)) +
    geom_histogram(binwidth = 0.5, fill = "skyblue", color = "black") +
    ggtitle(sprintf("v%d = (%.2f, %.2f)", i, vs[[i]][1], vs[[i]][2])) +
    xlab(sprintf("Shapiro-Wilk test p-value: %.2f", shapiro.test(projected_samples)$p.value)) +
    theme(plot.title = element_text(hjust = 0.5))
})

# Arrange plots
plots <- grid.arrange(p1, do.call(grid.arrange, histograms), ncol=2, widths = c(2, 1))

# Show plots
print(plots)

File: week2/ex16.py
import numpy as np
import matplotlib.pyplot as plt
from numpy.random import multivariate_normal
from scipy.stats import shapiro

rho = .5
cov = np.array([[1,rho], [rho, 1]])
mean = np.zeros(2)
N_samples = 1000
samples = multivariate_normal( mean, cov,N_samples)

# 3 different projections
v1 = np.array([1,0]) # x-axis
v2 = np.array([0,1]) # y-axis
v3 = np.array([1,1]); v3 = v3/np.linalg.norm(v3)
vs = [v1,v2,v3]



# scatterplot of samples
fig,ax = plt.subplots(figsize = (5,5))
ax.scatter(samples[:,0], samples[:,1]); 
fig.suptitle('Gaussian random samples')
ax.set_title('$\\rho$ = {}'.format(rho))

# 3 histograms of different projections
fig, axs = plt.subplots(figsize = (16,5), ncols = 3)
fig.suptitle('Histograms of projected data')
for i, (v, ax) in enumerate(zip(vs,axs)):
    projected_samples = np.array([ np.dot(x, v) for x in samples])
    ax.hist(projected_samples)
    ax.set_title('$v_{}$ = ({}, {})'.format(i+1, v[0].round(2), v[1].round(2)))
    ax.set_xlabel('Shapiro-Wilk test p-value: {:.2f}'.format(shapiro(projected_samples).pvalue))
    
plt.show()

